tests/cxx/instance_preds/Shape.cpp

```diff
@@ -0,0 +1,237 @@
+class Pos
+{
+  int m_x;
+  int m_y;
+
+public:
+  /*@ 
+    predicate valid(int x, int y) =
+      this->m_x |-> x &*& this->m_y |-> y;
+  @*/
+
+  Pos(int x, int y) : m_x(x), m_y(y) 
+  //@ requires true;
+  //@ ensures valid(x, y);
+  {
+    //@ close valid(x, y);
+  }
+
+  Pos() : Pos(0, 0)
+  //@ requires true;
+  //@ ensures valid(0, 0);
+  {}
+
+  Pos(const Pos &pos) : Pos(pos.getX(), pos.getY())
+  //@ requires pos.valid(?x, ?y);
+  //@ ensures valid(x, y) &*& pos.valid(x, y);
+  {}
+
+  ~Pos()
+  //@ requires valid(_, _);
+  //@ ensures true;
+  {
+    //@ open valid(_, _);
+  }
+
+  int getX() const
+  //@ requires valid(?x, ?y);
+  //@ ensures valid(x, y) &*& result == x;
+  {
+    //@ open valid(x, y);
+    return m_x;
+    //@ close valid(x, y);
+  }
+
+  int getY() const
+  //@ requires valid(?x, ?y);
+  //@ ensures valid(x, y) &*& result == y;
+  {
+    //@ open valid(x, y);
+    return m_y;
+    //@ close valid(x, y);
+  }
+
+  void setX(int x)
+  //@ requires valid(_, ?y);
+  //@ ensures valid(x, y);
+  {
+    //@ open valid(_, y);
+    m_x = x;
+    //@ close valid(x, y);
+  }
+
+  void setY(int y)
+  //@ requires valid(?x, _);
+  //@ ensures valid(x, y);
+  {
+    //@ open valid(x, _);
+    m_y = y;
+    //@ close valid(x, y);
+  }
+
+  void moveTo(int x, int y)
+  //@ requires valid(_, _);
+  //@ ensures valid(x, y);
+  {
+    setX(x);
+    setY(y);
+  }
+
+  void moveTo(const Pos &p)
+  //@ requires valid(_, _) &*& p.valid(?x, ?y);
+  //@ ensures valid(x, y) &*& p.valid(x, y);
+  {
+    int px = p.getX();
+    int py = p.getY();
+    moveTo(px, py);
+  }
+};
+
+class Shape
+{
+  Pos m_pos;
+
+public:
+  /*@
+    predicate valid(int x, int y) =
+      struct_Pos_padding(&this->m_pos) &*&
+      (&this->m_pos)->valid(x, y);
+  @*/
+
+  Shape(const Pos &pos) : m_pos(pos)
+  //@ requires pos.valid(?x, ?y);
+  //@ ensures pos.valid(x, y) &*& valid(x, y) &*& Shape_vtype(this, thisType);
+  {
+    //@ close valid(x, y);
+  }
+
+  virtual ~Shape()
+  //@ requires valid(_, _) &*& Shape_vtype(this, thisType);
+  //@ ensures true;
+  {
+    //@ open valid(_, _);
+  }
+
+  virtual int calcArea() const = 0;
+  //@ requires valid(?x, ?y);
+  //@ ensures valid(x, y) &*& result >= 0;
+};
+
+class Square : public Shape
+{
+  int m_width;
+
+public:
+  /*@
+    predicate valid(int x, int y) =
+      this->valid(&typeid(Shape))(x, y) &*&
+      this->m_width |-> ?w &*&
+      w >= 0 &*&
+      Square_bases_constructed(this);
+  @*/
+
+  Square(const Pos &pos, int width) : Shape(pos), m_width(width)
+  //@ requires pos.valid(?x, ?y) &*& width >= 0;
+  //@ ensures valid(x, y) &*& pos.valid(x, y) &*& Square_vtype(this, thisType);
+  {
+    //@ close valid(x, y);
+  }
+
+  ~Square()
+  //@ requires valid(_, _) &*& Square_vtype(this, thisType);
```

Comment:
I guess this destructor overrides the Shape destructor. Do you check spec compatibility? This succeeds here because `Square_vtype` is auto-opened to `Shape_vtype`? But what if Square has another polymorphic superclass as well? At first sight, it seems that virtual destruction of objects with multiple polymorphic superclasses is tricky because the `vtype` chunk of the other base class is not available? But we should postpone this issue until later if possible.

Reply:
Indeed, virtual destruction of objects with multiple polymorphic base classes would not succeed because the virtual destructor's spec in the derived class is checked against all virtual base destructor specs. It is (currently) also not supported to override multiple instance predicates from base classes in a derived class.

Reply:
👍 Fine like this for now.